src/transform.rs

```diff
@@ -0,0 +1,365 @@
+        let children = [&mut self.all_of, &mut self.any_of, &mut self.one_of]
+            .into_iter()
+            .filter_map(|x| x.as_mut())
+            .flat_map(|schemas| schemas.iter_mut())
+            .chain(self.not.as_deref_mut())
+            .chain(self.if_schema.as_deref_mut())
+            .chain(self.then_schema.as_deref_mut())
+            .chain(self.else_schema.as_deref_mut());
+
+        for child in children {
+            child.post_visit(visitor);
+        }
+    }
+}
+impl PostVisit for ArrayValidation {
+    fn post_visit<V: VisitorMut>(&mut self, visitor: &mut V) {
+        // SingleOrVec doesn't give a way to get a mutable iterator, so this was
+        // the easiest way I found to iterate over it.
+        if let Some(items) = &mut self.items {
+            match items {
+                schemars::schema::SingleOrVec::Single(x) => x.post_visit(visitor),
+                schemars::schema::SingleOrVec::Vec(vec) => {
+                    for x in vec {
+                        x.post_visit(visitor)
+                    }
+                }
+            }
+        }
+        let children = self.additional_items.iter_mut().chain(&mut self.contains);
+        for child in children {
+            child.post_visit(visitor);
+        }
+    }
+}
+
+impl PostVisit for ObjectValidation {
+    fn post_visit<V: VisitorMut>(&mut self, visitor: &mut V) {
+        let children = self
+            .properties
+            .values_mut()
+            .chain(self.pattern_properties.values_mut())
+            .chain(self.additional_properties.as_deref_mut())
+            .chain(self.property_names.as_deref_mut());
+
+        for child in children {
+            child.post_visit(visitor);
+        }
+    }
+}
+
+struct UnionType<'a> {
+    root: &'a RootSchema,
+}
+
+impl VisitorMut for UnionType<'_> {
+    fn visit_object(&mut self, obj: &mut SchemaObject) {
+        if obj.instance_type.is_some() {
+            return;
+        }
+
+        if let Some(subschemas) = &mut obj.subschemas {
+            match (&subschemas.any_of, &subschemas.one_of) {
+                (None, Some(any_of)) | (Some(any_of), None) => {
+                    let Some(plain_types) = any_of
+                        .iter()
+                        .map(|s| plain_schema_types(s, self.root))
+                        .collect::<Option<Vec<_>>>()
+                    else {
+                        return;
+                    };
+
+                    let Some(mut plain_types) = plain_types
+                        .into_iter()
+                        .map(|tys| match tys {
+                            SingleOrVec::Single(t) => Some(*t),
+                            SingleOrVec::Vec(_) => None,
+                        })
+                        .collect::<Option<Vec<_>>>()
+                    else {
+                        return;
+                    };
+
+                    // Technically, if this is `one_of` then we should filter out any repeated types.
```

Comment:
Why don't we do it? It does seem like this is changing the semantics of `one_of`, although this sounds like a "nobody in their right mind would do that" case.

Reply:
A combination of laziness and "nobody in their right mind would do that". I've fixed it now (and print a warning if they aren't in their right mind).